File: Exercise3/Noise.cpp
#include "opencv2/core/core.hpp"
#include "opencv2/highgui/highgui.hpp"
#include "opencv2/imgproc/imgproc.hpp"
#include "iostream"
 
 // Author Rishab Shah

using namespace cv;
using namespace std;

void Add_salt_pepper_Noise(Mat &srcArr, float pa, float pb );
void Add_gaussian_Noise(Mat &srcArr,double mean,double sigma);
 
int main(int argc, char** argv )
{
    Mat image;
    image = imread(argv[1], 0);
    namedWindow( "Original image", CV_WINDOW_AUTOSIZE );
    imshow( "Original image", image);

    Mat noise_img = image.clone(); 
    float mean = 0;			
    float sigma = 50;		//Not the variance
    Add_gaussian_Noise(noise_img, mean, sigma);
    imshow( "Gaussian Noise", noise_img);

    Mat noise_dst = noise_img.clone();
    blur(noise_dst,noise_dst,Size(3,3));
    imshow( "Box filter", noise_dst);

    Mat noise_dst1 = noise_img.clone();
    GaussianBlur(noise_dst1, noise_dst1, Size(3,3),1.5);
    imshow( "Gaussian filter", noise_dst1);
    
    Mat noise_dst2 = noise_img.clone();
    medianBlur(noise_dst2,noise_dst2,3);
    imshow( "Median filter", noise_dst2);
    
    Mat noise_img2 = image.clone(); 
    float pa=0.01;			//Controls the amount of black spots in the noise
    float pb=0.01;			//Controls the amount of white spots in the noise
    Add_salt_pepper_Noise(noise_img2, pa, pb); 
    imshow( "Salt and Pepper Noise", noise_img2);

    Mat noise_dst3 = noise_img2.clone();
    blur(noise_dst3,noise_dst3,Size(3,3));
    imshow( "Box filter", noise_dst3);

    Mat noise_dst4 = noise_img2.clone();
    GaussianBlur(noise_dst4,noise_dst4,Size(3,3),1.5);
    imshow( "Gaussian filter", noise_dst4);
    
    Mat noise_dst5 = noise_img2.clone();
    medianBlur(noise_dst5,noise_dst5,3);
    imshow( "Median filter", noise_dst5);
    return 0;
}

void Add_salt_pepper_Noise(Mat &srcArr, float pa, float pb )

{    RNG rng; // rand number generate
    int amount1=srcArr.rows*srcArr.cols*pa; //Need to run through all channels
    int amount2=srcArr.rows*srcArr.cols*pb;
    for(int counter=0; counter<amount1; ++counter)
    {
     srcArr.at<uchar>(rng.uniform( 0,srcArr.rows), rng.uniform(0, srcArr.cols)) =0;

    }
     for (int counter=0; counter<amount2; ++counter)
     {
     srcArr.at<uchar>(rng.uniform(0,srcArr.rows), rng.uniform(0,srcArr.cols)) = 255;
     }
}


void Add_gaussian_Noise(Mat &srcArr,double mean,double sigma)
{
    Mat NoiseArr = srcArr.clone();
    RNG rng;
    rng.fill(NoiseArr, RNG::NORMAL, mean,sigma);  
    add(srcArr, NoiseArr, srcArr);   
}

File: Exercise3/noise.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Sun Nov  5 20:42:28 2017

@author: zhangzhexi
"""
import cv2
import numpy as np
import random

def Add_salt_pepper_Noise(image,pa,pb):
    row,col,ch=image.shape
    amount1=row*col*pa
    amount2=row*col*pb
    for i in range(int(amount1)):
        image[int(random.uniform(0,row))][int(random.uniform(0,col))]=0
    for i in range(int(amount2)):
        image[int(random.uniform(0,row))][int(random.uniform(0,col))]=255

def Add_gaussian_Noise(src,mean,sigma):
    noiseArr = src.copy()
    noiseArr = np.random.normal(mean,sigma,src.shape)
    np.add(src,noiseArr,src,casting="unsafe")
    return;

def main():
    img=cv2.imread('Test_images/Lenna.png')
    cv2.namedWindow('Original image')
    cv2.imshow('Original',img)
    cv2.imwrite('Original.jpg',img)
    
    noise_img=img.copy()
    Add_gaussian_Noise(noise_img,0,50)
    cv2.imshow('Gaussian Noise',noise_img)
    cv2.imwrite('Gaussian_Noise.jpg',noise_img)
    
    noise_dst=noise_img.copy()
    cv2.blur(noise_dst,(3,3))
    cv2.imshow('Box filter',noise_dst)
    cv2.imwrite('Box_filter.jpg',noise_dst)
    
    noise_dst1=noise_img.copy()
    cv2.GaussianBlur(noise_dst1,(3,3),1.5)
    cv2.imshow('GaussianBlur filter',noise_dst1)
    cv2.imwrite('Gaussian_Blur_filter.jpg',noise_dst1)
    
    noise_dst2=noise_img.copy()
    cv2.medianBlur(noise_dst2,3)
    cv2.imshow('Median filter',noise_dst2)
    cv2.imwrite('Median_filter.jpg',noise_dst2)
    
    noise_img2=img.copy()
    pa=0.01     #Controls the amount of black spots in the noise
    pb=0.01     #Controls the amount of white spots in the noise
    Add_salt_pepper_Noise(noise_img2,pa,pb)
    cv2.imshow("Salt and Peper Noise", noise_img2)
    cv2.imwrite("Salt_and_Peper_Noise.jpg", noise_img2)
    
    noise_dst3=noise_img2.copy()
    cv2.blur(noise_dst3,(3,3))
    cv2.imshow('Box filter2',noise_dst3)
    cv2.imwrite('Box_filter2.jpg',noise_dst3)
    
    noise_dst4=noise_img2.copy()
    cv2.GaussianBlur(noise_dst4,(3,3),1.5)
    cv2.imshow('Gaussian Blur filter2',noise_dst4)
    cv2.imwrite('Gaussian_Blur_filter2.jpg',noise_dst4)
    
    noise_dst5=noise_img2.copy()
    cv2.medianBlur(noise_dst5,3)
    cv2.imshow('Median filter2',noise_dst5)
    cv2.imwrite('Median_filter2.jpg',noise_dst5)
    
    cv2.waitKey(0)
    cv2.destroyAllWindows()
    
if __name__ == '__main__':
    main()
    
